File: dp/coin_change.cpp
// leetcode 322. Coin Change
// medium
// dp[amount] = dp[amount - coin_val] + 1 for coin_val in coins

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, -1);
        dp[0] = 0;
        for(int i = 0; i < dp.size(); i++){
            for(int c : coins){
                if(i - c >= 0 and dp[i-c] != -1){
                    int temp = dp[i-c] + 1;
                    if(dp[i] == -1){
                        dp[i] = temp;
                    }
                    else {
                        dp[i] = min(dp[i], temp);
                    }
                }
            }
        }
        return dp[amount];
    }
};

File: dp/coin_change.py
# 322. Coin Change
# medium

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        coins.sort(reverse=True)
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            for coin in coins:
                if coin > i:
                    continue
                leftOver = i - coin
                if dp[leftOver] != -1:
                    dp[i] = min(dp[leftOver] + 1, dp[i])
        return dp[amount] if dp[amount] <= amount else -1